[SYCL] Enable level zero v2 in DPC++

omarahmed1111 opened the pull request:
This makes a simple way to use the experimental level_zero_v2 in dpc++. Fixes https://github.com/intel/llvm/issues/16613 .

pbalcer commented:
My opinion is that we should build and ship both adapters by default, and make it togglable which one is used at runtime through an env variable (like I suggested [here](https://github.com/oneapi-src/unified-runtime/pull/2571#issuecomment-2595159134)).
Having both is often useful for performance comparisons, or to simply to have a fallback if one of the adapters is not behaving correctly.

omarahmed1111 commented:
> My opinion is that we should build and ship both adapters by default, and make it togglable which one is used at runtime through an env variable (like I suggested [here](https://github.com/oneapi-src/unified-runtime/pull/2571#issuecomment-2595159134)). Having both is often useful for performance comparisons, or simply to have a fallback if one of the adapters is not behaving correctly.

Good point! changed that.

jbrodman commented:
By default, I think we should only build 1 adapter. We don't want to confuse users by enumerating the same device even more times than we already do (OpenCL vs Level Zero). I think it would be nice to support building both and toggling for the reasons stated above, but I'd also like the option to ONLY build v1 or v2.

pbalcer commented:
> By default, I think we should only build 1 adapter. We don't want to confuse users by enumerating the same device even more times than we already do (OpenCL vs Level Zero). I think it would be nice to support building both and toggling for the reasons stated above, but I'd also like the option to ONLY build v1 or v2.

The current proposal, and, from what I can tell, the implementation, is that both adapters will be physically present in the libs directory, however, only one of them will be active at a time. There will be a new environment variable (`UR_ADAPTER_LEVEL_ZERO_V2=1`, although, long-term, I'd have preferred for this to be an option for `ONEAPI_DEVICE_SELECTOR`), that users will be able to set to toggle which adapter is used. This ensures that L0 devices are not enumerated twice, but makes it very simple for users to experiment with the new adapter.

Having to build SYCL from scratch with a special option might be challenging for developers of higher-level frameworks or applications that usually use a prebuilt compiler.

sarnex commented on buildbot/configure.py:
```diff
@@ -68,6 +68,7 @@ def do_configure(args):
 
     if sys.platform != "darwin":
         sycl_enabled_backends.append("level_zero")
+        sycl_enabled_backends.append("level_zero_v2")
```
Couple of comments

1) Can you explain what L0 V2 is and how it releases to the current L0/L0 adapter

2) I agree with James, we shouldn't build two plugins by default. If we have some configure switch to build two, that's fine.

pbalcer replied:
1. [README](https://github.com/oneapi-src/unified-runtime/tree/main/source/adapters/level_zero/v2). In short, it's a new L0 adapter with re-implemented performance-critical APIs. See [here](https://oneapi-src.github.io/unified-runtime/benchmark_results.html?suites=Compute+Benchmarks&regex=SubmitKernel) for performance validation results.
2. Why though? Users won't be affected by the inclusion of the two adapters. And not including it by default would mean that it's going to be much more difficult for application and framework developers to experiment with the new adapter, which is the whole point of including it in SYCL right now in the first place. In practice, it will likely mean we might need to provide the users that want to use it an alternative prebuilt compiler package.

@igchor ping

aelovikov-intel replied:
Out of curiosity, why "v2" instead of incremental refactoring?

omarahmed1111 replied:
I think @pbalcer might have a better answer than me for this questions.

omarahmed1111 commented on sycl/cmake/modules/FetchUnifiedRuntime.cmake:
```diff
@@ -36,6 +36,10 @@ set(UR_ENABLE_TRACING ON)
 endif()
+if("level_zero_v2" IN_LIST SYCL_ENABLE_BACKENDS)
+  set(UR_BUILD_ADAPTER_L0_V2 ON)
+  set(ENV{UR_ADAPTER_LEVEL_ZERO_V2} ON)
```
I tried to think in an easy way to do the second option and make the compile option consistent in defining the env variable, but I think we will always need the user to define the `UR_ADAPTER_LEVEL_ZERO_V2`, I guess we could just depend on the user to define it, or we could define another dpc++ env variable that the user could define if that would make it more clear. 

The first option seems the simpler one, but with the probability of reporting still 2 l0 adapters. 

I am not sure which one is better, but I think we could just for simplicity go with the first suggestion of just registering the two `l0/l0_v2` adapters, and let dpc++ control that with the compile option.

pbalcer replied:
Makes sense, I agree. I already approved the UR change. Thanks.

igchor commented:
@omarahmed1111 will this approach enable us to mark specific e2e tests with XFAIL for v2 only, or would that require some extra changes?

omarahmed1111 commented:
> @omarahmed1111 will this approach enable us to mark specific e2e tests with XFAIL for v2 only, or would that require some extra changes?

I guess that will need extra changes to lit tests config to be able to detect `level_zero_v2` as a separate backend that could be marked as xfail, not sure if we need also to make sycl-rt detect level_zero_v2 as a valid backend for that to work.

pbalcer commented:
> > @omarahmed1111 will this approach enable us to mark specific e2e tests with XFAIL for v2 only, or would that require some extra changes?
> 
> I guess that will need extra changes to lit tests config to be able to detect `level_zero_v2` as a separate backend that could be marked as xfail, not sure if we need also to make sycl-rt detect level_zero_v2 as a valid backend for that to work.

Let's leave that for a separate PR. This one isn't adding any new CI job for v2 anyway.

omarahmed1111 commented:
@intel/llvm-reviewers-runtime please review, Thanks!

againull commented on sycl/CMakeLists.txt:
```diff
@@ -176,7 +176,7 @@ if ("opencl" IN_LIST SYCL_ENABLE_BACKENDS)
 endif()
 if ("level_zero" IN_LIST SYCL_ENABLE_BACKENDS)
-  set(SYCL_BUILD_BACKENDLEVEL_ZERO ON)
+  set(SYCL_BUILD_BACKEND_LEVEL_ZERO ON)
```
If I am not mistaken the only usage of this variable is in LIT to enable level_zero feature. I wonder how it even worked with this typo and whether we were just skipping tests which had level_zero requirement.

EwanC commented on buildbot/configure.py:
```diff
@@ -327,6 +330,9 @@ def main():
         help="choose hardware platform for HIP backend",
     )
+    parser.add_argument(
+        "--level_zero_v2", action="store_true", help="Enable SYCL level_zero_v2"
```
Slight preference to not have variable names in user help
```suggestion
        "--level_zero_v2", action="store_true", help="Enable SYCL Level Zero V2"
```

omarahmed1111 replied:
Good point, changed that

kbenzie reviewed (approved):
UR LGTM

omarahmed1111 commented:
@intel/dpcpp-devops-reviewers Please review, Thanks!

omarahmed1111 commented:
@intel/dpcpp-devops-reviewers Gentle ping on this!

pbalcer commented:
@intel/dpcpp-devops-reviewers please review. This is blocking our progress on properly enabling v2 in tests and the compiler packages.
Thanks!

igchor commented:
> Please add links to the in-tree documentation describing rationale and short/mid/long-term plans for this.

I believe @pbalcer explained the rationale in his previous comment. As for the plans, we want to provide v2 as an experimental adapter in the 2025.2 release to gather feedback from the users and, if needed, add missing features (v2 does not implement support for all env variables that are used to tweak the legacy adapter). In 2026.0, we plan to make the v2 adapter the default one for L0.

I guess we can add this information to https://github.com/intel/llvm/blob/sycl/unified-runtime/source/adapters/level_zero/v2/README.md

aelovikov-intel commented:
> I believe @pbalcer explained the rationale in his previous comment.

I didn't ask for the explanation itself, but to put the link to it as a comment in the code, so that it would be immediately available for anyone reading the code in future.

aelovikov-intel reviewed (approved):
> Please add links to the in-tree documentation describing rationale and short/mid/long-term plans for this

You've only added the former and not the latter... Also, having links to a comments thread in PR review is weird at the very least.

Anyhow, someone else gave you an approval for the SYCL RT, so I'm only LGTM'ing the configure.py change following those pre-approved changes.

omarahmed1111 commented:
> > Please add links to the in-tree documentation describing rationale and short/mid/long-term plans for this
> 
> You've only added the former and not the latter... Also, having links to a comments thread in PR review is weird at the very least.
> 
> Anyhow, someone else gave you an approval for the SYCL RT, so I'm only LGTM'ing the configure.py change following those pre-approved changes.

Apologies, I misunderstood that, I modified the comment to reference the page @igchor mentioned and added a small comment about the timeline. I didn't want to add more than that as I think L0_V2 team could reference this document explanations better. So, I just will leave this for the functionality and this small comment.

omarahmed1111 commented:
@intel/llvm-gatekeepers Please merge, Thanks!
